feat: Add function to list tables that match a source spec

hermanschaaf opened the pull request:
This adds a new `GetTablesForSpec` function to the `SourceClient` that filters the list of tables to the ones selected by the given Source Spec. It is fully backwards-compatible, and I decided to keep the changeset small by adding an optional parameter to the existing `GetTables` proto rpc.

This will be used to implement https://github.com/cloudquery/cloudquery/issues/4715

yevgenypats reviewed (changes requested):
Will it make sense to introduce another method for that? Otherwise the client can think this is supported while it doesn't and if it will be a new method it can distinguish that. Not sure if this is needed as it can also work that way. but just bringing this up for discussion.

hermanschaaf commented:
@yevgenypats Hmm yeah, good point. We don't need to know right now, but I can see it potentially being important/useful to avoid bugs in the future. I'm fine with adding an extra method if you're happy with that too.

hermanschaaf commented:
@yevgenypats I've added the method, and updated the corresponding PR here https://github.com/cloudquery/cloudquery/pull/5098 to make sure it all works together. Here's an updated example using the new code:

```
➜  playground ../cloudquery/cli/cli sync aws.yml
Loading spec(s) from aws.yml
Source aws will migrate and sync 23 tables.
Starting migration for: aws -> [postgresql]
Migration completed successfully.
Starting sync for: aws -> [postgresql]
Sync completed successfully. Resources: 5, Errors: 1, Panics: 0, Time: 3s
```

And if a new CLI is run against an old version of the AWS plugin:

```
➜  playground ../cloudquery/cli/cli sync aws.yml
Loading spec(s) from aws.yml
Starting migration for: aws -> [postgresql]
Migration completed successfully.
Starting sync for: aws -> [postgresql]
Sync completed successfully. Resources: 5, Errors: 1, Panics: 0, Time: 4s
```

Please take another look 🙇

yevgenypats commented on clients/source.go:
```diff
@@ -24,6 +24,8 @@ import (
 	"google.golang.org/grpc/status"
 )
 
+var ErrMethodNotSupported = errors.New("method not supported")
```
I don't think we should do that because but rather we need to use the grpc code directly. The CLI can have a helper function that check if a grpc is unimplemented and return some error message but the client shouldn't introduce a new error.

hermanschaaf replied:
Maybe I misunderstand the purpose of the client library, but my thinking is that it should abstract away the gRPC calls so that users of this library don't have to concern themselves with the underlying gRPC layer?

hermanschaaf replied:
Oh yeah, it's probably also worth mentioning that the gRPC errors don't support `errors.As`, otherwise it would have been okay to wrap the error before returning it: https://github.com/grpc/grpc-go/issues/2934. But now, since it's not supported, 

But as things stand, it's pretty tricky for callers of the client library to discover when it's not supported in a reliable way. Probably the best would be to look for a specific string in the error message.

yevgenypats replied:
I see. ok I guess it's a nit. not sure right now what would be the best solution so let's skip it. mostly the things that worries me that it is now inconsistent because then we need to put it in all calls.

yevgenypats reviewed (changes requested):
Looks good. One comment on error handling and I think one more "e2e" test is missing here https://github.com/cloudquery/plugin-sdk/blob/main/serve/source_test.go

hermanschaaf commented:
@yevgenypats Thanks for the review! Updated with the following changes:
  - added `IsUnimplemented` helper function to help determine when the underlying method was unimplemented by the server
  - added an e2e test for `GetTablesForSpec`
  - no longer returning a custom error, but instead wrapping the grpc one
